documentation/high-avail/ha-api-hld.md

```diff
@@ -0,0 +1,617 @@
+    SAI_DASH_HA_ROLE_SWITCHING_TO_ACTIVE,
+} sai_dash_ha_role_t;
+```
+
+### 4.3. Flow table
+
+HA uses the DASH flow table to achieve the flow state manipulation. Since the flow table already provides the CRUD operations, we don't need any extra APIs from flow table.
+
+For more information, please refer to DASH flow API documentation.
+
+### 4.4. Flow
+
+To support HA, each flow contains the following SAI attributes:
+
+| Attribute name | Type | Description |
+| -------------- | ---- | ----------- |
+| SAI_FLOW_ATTR_FLOW_VERSION | `sai_uint32_t` | The flow version. |
+| SAI_FLOW_ATTR_FLOW_SYNC_STATE | `sai_dash_ha_flow_sync_state_t` | The flow sync state. |
+
+The flow sync state is defined as below:
+
+```c
+typedef enum _sai_dash_ha_flow_sync_state_t
+{
+    SAI_DASH_HA_FLOW_SYNC_STATE_FLOW_MISS,
+    SAI_DASH_HA_FLOW_SYNC_STATE_FLOW_CREATED,
+    SAI_DASH_HA_FLOW_SYNC_STATE_FLOW_SYNCED,
+    SAI_DASH_HA_FLOW_SYNC_STATE_FLOW_PENDING_DELETE,
+    SAI_DASH_HA_FLOW_SYNC_STATE_FLOW_PENDING_RESIMULATION
+} sai_dash_ha_flow_sync_state_t;
+```
+
+The flow sync state is a small state machine that represents if the flow is synched or not, so we can make the packet processing decision accordingly to achieve HA.
+
+```mermaid
+stateDiagram-v2
+    M: FLOW_MISS
+    C: FLOW_CREATED
+    S: FLOW_SYNCED
+    D: FLOW_PENDING_DELETE
+    R: FLOW_PENDING_RESIMULATION
+
+    M --> C: Flow created
+    M --> S: Flow sync request<br>received on<br>standby node
+
+    C --> S: Flow sync ack received
+    S --> R: Flow resimulation requested
+    R --> C: Flow updated
+
+    C --> D: Flow deleted
+    S --> D: Flow deleted
+    R --> D: Flow deleted
+
+    D --> M: Flow sync ack received
+```
+
+For more information, please refer to DASH flow API documentation.
+
+### 4.5. ENI
+
+To provide the ENI-level HA control, each ENI will have the following SAI attributes:
+
+| Attribute name | Type | Description |
+| -------------- | ---- | ----------- |
+| SAI_ENI_ATTR_HA_SCOPE_ID | `sai_object_id_t` | The HA scope ID of the ENI. |
+
+### 4.6. Event notifications
+
+To receive the HA related updates from the DASH implementation, the following SAI notification attributes are added on the switch object:
+
+| Attribute name | Type | Description |
+| -------------- | ---- | ----------- |
+| SAI_SWITCH_ATTR_HA_SET_EVENT_NOTIFY | `sai_ha_set_event_notification_fn` | The callback function for receiving events on the HA set. |
+| SAI_SWITCH_ATTR_HA_SCOPE_EVENT_NOTIFY | `sai_ha_scope_event_notification_fn` | The callback function for receiving events on the HA scope. |
+
+#### 4.6.1. HA set event notifications
+
+Whenever a HA set state is changed, it will be reported back via HA set event notification, such as data plane channel goes down. The detailed definition is shown as below:
+
+```c
+/**
+ * @brief HA set event type
+ */
+typedef enum _sai_ha_set_event_t
+{
+    /** Any HA set state is changed, such as data plane channel goes down. */
+    SAI_HA_SET_STATE_CHANGED,
```

Comment:
This enum should have the possible states to remove the need to read them from the attribute.

Reply:
ah, yea. i am mostly following what other SAI events is doing, but please correct me if my understanding is not right.

yes, the event enum doesn't contain the detailed state, but we don't need to read them from HA set attribute using query. the reason is the changed attributes will be reported as attr_count and attr in the sai_ha_set_event_data, hence we can directly fetch them there.

I found this approach being a repetitive pattern in existing SAI APIs, so i am mostly following this pattern. It does simplify the event data definition and won't cause state explosion on the enum value.

Reply:
Notification data doesn't carry attributes but the type that attribute has. See https://github.com/opencomputeproject/SAI/blob/master/inc/saiport.h#L90C5-L90C28 for reference.

Reply:
yea, i know ACL and Port notification is doing it, but it might not be extendable, because we need to change the API whenever a new field needs to be added.

on the otherhand, [FDB](https://github.com/opencomputeproject/SAI/blob/3fe490d836be6ed1d7c4a8763d5e33053f9e957a/inc/saifdb.h#L346), [packet event](https://github.com/opencomputeproject/SAI/blob/3fe490d836be6ed1d7c4a8763d5e33053f9e957a/inc/saihostif.h#L1417) are using attributes directly, which seems to be a more flexiable approach, but i guess the thing I am missing here is the bar, FDB needs to report all its attributes, hence designed like that.

so no problem at all, let me get the event updated.

Reply:
FDB attribute is used to indicate a bridge ID and an entry type - https://github.com/sonic-net/sonic-swss/blob/master/orchagent/fdborch.cpp#L1054, not the state. State is passed as part of the notification data separately. Why would it be harder to extend if both are using the same type?